webapp/tests/test_render_glyph.py

```diff
@@ -0,0 +1,1002 @@
+
+    def test_safeArgs_number_not_list(self):
+      with self.assertRaises(TypeError):
+        glyph.safeArgs(1)
+
+    def test_safeArgs_all_strings(self):
+      with self.assertRaises(TypeError):
+        list(glyph.safeArgs(['a','b','c']))
+
+    def test_safeArgs_list_with_string(self):
+      with self.assertRaises(TypeError):
+        self.assertEqual(list(glyph.safeArgs(['a',5.1,6])), ['a',5.1,6])
+
+    def test_safeArgs_empty_string(self):
+      self.assertEqual(list(glyph.safeArgs('')), [])
+
+    def test_safeArgs_list_all_numbers(self):
+      self.assertEqual(list(glyph.safeArgs([1,5.1,6,])), [1,5.1,6])
+
+    def test_safeArgs_list_with_nan(self):
+      self.assertEqual(list(glyph.safeArgs([1,5.1,6,float('Nan')])), [1,5.1,6])
+
+    def test_safeArgs_list_with_inf(self):
+      self.assertEqual(list(glyph.safeArgs([1,5.1,6,float('inf')])), [1,5.1,6])
+
+    def test_safeArgs_list_with_None(self):
+      self.assertEqual(list(glyph.safeArgs([1,5.1,6,None])), [1,5.1,6])
+
+    #
+    # Testing safeMin()
+    #
+    def test_safeMin_None(self):
+      with self.assertRaises(TypeError):
+        glyph.safeMin(None)
+
+    def test_safeMin_number_not_list(self):
+      with self.assertRaises(TypeError):
+        glyph.safeMin(1)
+
+    def test_safeMin_empty_string(self):
+      self.assertEqual(glyph.safeMin(''), None)
+
+    def test_safeMin_list_with_string(self):
+      with self.assertRaises(TypeError):
+        glyph.safeMin(['a',10,30])
+
+    def test_safeMin_list_all_numbers(self):
+      self.assertEqual(glyph.safeMin([1,5.1,6,]), 1)
+
+    def test_safeMin_list_with_nan(self):
+      self.assertEqual(glyph.safeMin([1,5.1,6,float('Nan')]), 1)
+
+    #
+    # Testing safeMax()
+    #
+    def test_safeMax_None(self):
+      with self.assertRaises(TypeError):
+        glyph.safeMax(None)
+
+    def test_safeMax_number_not_list(self):
+      with self.assertRaises(TypeError):
+        glyph.safeMax(1)
+
+    def test_safeMax_empty_string(self):
+      self.assertEqual(glyph.safeMax(''), None)
+
+    def test_safeMax_list_with_string(self):
+      with self.assertRaises(TypeError):
+        glyph.safeMax(['a',10,30])
+
+    def test_safeMax_list_all_numbers(self):
+      self.assertEqual(glyph.safeMax([1,5.1,6,]), 6)
+
+    def test_safeMax_list_with_nan(self):
+      self.assertEqual(glyph.safeMax([1,5.1,6,float('Nan')]), 6)
+
+    #
+    # Testing safeSum()
+    #
+    def test_safeSum_None(self):
+      with self.assertRaises(TypeError):
+        glyph.safeSum(None)
+
+    def test_safeSum_number_not_list(self):
+      with self.assertRaises(TypeError):
+        glyph.safeSum(1)
+
+    def test_safeSum_empty_string(self):
+      self.assertEqual(glyph.safeSum(''), 0)
+
+    def test_safeSum_list_with_string(self):
+      with self.assertRaises(TypeError):
+        glyph.safeSum(['a',10,30])
+
+    def test_safeSum_list_all_numbers(self):
+      self.assertEqual(glyph.safeSum([1,5.1,6,]), 12.1)
```

Comment:
Tests involving floating-point numbers are often subject to floating-point rounding errors, because `assertEqual()` insists that the numbers are identical, whereas they might, say, differ in the 13th decimal place due to rounding. `assertAlmostEqual()` is more robust in situations like this.